src/Avalonia.Controls/Primitives/VisualLayerManager.cs

```diff
@@ -0,0 +1,97 @@
+using System.Collections.Generic;
+using Avalonia.LogicalTree;
+using Avalonia.Styling;
+
+namespace Avalonia.Controls.Primitives
+{
+    public class VisualLayerManager : Decorator
+    {
+        private const int AdornerZIndex = int.MaxValue - 100;
+        private const int OverlayZIndex = int.MaxValue - 99;
+
+        private bool _isAttachedToLogicalTree;
+        private IStyleHost _styleHost;
+        public bool IsPopup { get; set; }
```

Comment:
Nit: Can we move this to be with other properties? It gets mistaken as a private field here.